examples/imagenet/alexnet.cc

```diff
@@ -0,0 +1,403 @@
+          }
+          break;
+        }
+      }
+      if (n_read == read_size) {
+        train_x.CopyData(prefetch_x);
+        train_y.CopyData(prefetch_y);
+      }
+      //LOG(INFO) << "train_x.L1(): " << train_x.L1();
+      timer.Tick();
+      th = data.AsyncLoadData(kTrain, binfile, read_size, &prefetch_x, &prefetch_y, &n_read);
+      if (n_read < read_size) continue;
+      CHECK_EQ(train_x.shape(0), train_y.shape(0));
+ //     train_x.ToDevice(device);
+//      train_y.ToDevice(device);
+      ttr.Tick();
+      auto ret = net.TrainOnBatch(epoch, train_x, train_y);
+      train_time += ttr.Elapsed();
+      loss += ret.first;
+      metric += ret.second;
+      b++;
+      //LOG(INFO) << "batch " << b << ", loss: " << ret.first;
+    }
+    if (fno % 20 == 0) {
+      //LOG(INFO) << "num of batch: " << b;
+      train_ch->Send("Epoch " + std::to_string(epoch) + ", training loss = " +
+                      std::to_string(loss / b) + ", accuracy = " +
+                      std::to_string(metric / b) + ", lr = " +
+                      std::to_string(lr)
+                      + ", time of loading " + std::to_string(read_size) + " images = "
+                      + std::to_string(load_time / b) + " ms, time of training (batchsize = "
+                      + std::to_string(read_size) + ") = " + std::to_string(train_time / b) + " ms.");
+      loss = 0.0f;
+      metric = 0.0f;
+      load_time = 0.0f;
+      train_time = 0.0f;
+      b = 0;
+    }
+  }
+}
+
+void TestOneEpoch(FeedForwardNet &net, ILSVRC &data,
+                                     std::shared_ptr<Device> device, int epoch,
+                                     string bin_folder, size_t num_test_images,
+                                     size_t read_size,
+                                     Channel *val_ch) {
+  float loss = 0.0f, metric = 0.0f;
+  float load_time = 0.0f, eval_time = 0.0f;
+  size_t n_read;
+  string binfile = bin_folder + "/test.bin";
+  Timer timer, tte;
+  Tensor prefetch_x, prefetch_y;
+  //prefetch_x.ToDevice(device);
+  //prefetch_y.ToDevice(device);
+  timer.Tick();
+  data.LoadData(kEval, binfile, read_size, &prefetch_x, &prefetch_y, &n_read);
+  load_time += timer.Elapsed();
+  Tensor test_x(prefetch_x.shape(), device);
+  Tensor test_y(prefetch_y.shape(), device, kInt);
+  int remain = (int)num_test_images - n_read;
+  //LOG(INFO) << "num of test images: " << remain;
+  CHECK_EQ(n_read, read_size);
+  //test_x.ResetLike(prefetch_x);
+  //test_y.ResetLike(prefetch_y);
+  std::thread th;
+  while (true) {
+    if (th.joinable()) {
+      th.join();
+      load_time += timer.Elapsed();
+      remain -= n_read;
+      if (remain < 0) break;
+      if (n_read < read_size) break;
+    }
+    //LOG(INFO) << "num of test images: " << remain;
+    test_x.CopyData(prefetch_x);
+    test_y.CopyData(prefetch_y);
+    timer.Tick();
+    th = data.AsyncLoadData(kEval, binfile, read_size, &prefetch_x, &prefetch_y, &n_read);
```

Comment:
if all test images are load together in one call, then no need to use AsyncLoadData.